### PR TITLE
[DEX-263] Updated Swagger UI Documentation Building

### PR DESCRIPTION
**Pull request by bluemellophone**
## Pull Request Overview

- Updated the way we download, build, and install Swagger UI web assets.  
  - This PR pulls inspiration from `scripts/build.frontend.sh` and Flask RESTX's implementation (https://github.com/python-restx/flask-restx/blob/master/tasks.py#L193) where we are downloading and building the Swagger UI components with NPM inside a Docker image.  
  - To achieve the NPM build, we define a new `docs/package.json` and link it into the build container similar to the frontend process. The build files are stored in `docs/swagger_static/` and then installed into `app/static/swagger-ui/`.  
- Updated the HTML visualization of Swagger UI JSON spec file, also pulling inspiration from Flask RESTX.  
  - We are now relying on the built-in flask_restx.apidoc implementation and are able to delete our route rules in `app/modules/frontend/views_documentation.py` (in fact, we can delete the entire file).
- Rebalanced the invoke tasks for dependency installation.
  - There is now a dedicated `dependencies.install-all-ui` and `dependencies.install-frontend-ui` to help standardize the usage.  The previous behavior for `dependencies.install` has been focused in the places it is needed to use `dependencies.install-python-dependencies` instead.
- Updated various bash scripts in `scripts/` including `scripts/clean.sh` to delete any added Node/NPM build caches in `docs/`
- Fixes an inherited error from `main` (and 3 failing tests) with `site-settings` and `tempfile.TemporaryDirectory`.
- Updated the Pull Request Template to drop `develop` in favor of `main`
- Miscellaneous clean-up, wording updates, and linting of random files. 

Screen Shot 2021-04-14 at 11 45 40 PM

---

**Review Notes**

Note 1 - ~~Looks like `pytest tests/modules/site_settings/test_resources.py::test_site_settings` is consistently failing with Python 3.7 only.~~ [Fixed]

Note 2 - Since we are building the Swagger UI static files inside a Docker image -- similar to how the frontend is built and linked into `dist-latest` -- we therefore cannot build the Swagger dependency when running inside an image.  Previously the installation required downloading a frozen version of Swagger UI and could be performed during the Dockerfile build process.  Instead, the updated implementation gets the latest version directly from NPM and builds them for our use inside a container, which predicates that we are not already inside a container unless we employ some sort of complex DnD setup (not advised).  This means that prior to launching a container-based deployment, both the Frontend and Swagger UI static files need to be built and installed.

All of the new Swagger UI static files are being installed to `app/static/swagger-ui/` and are referenced from the updated template `app/templates/swagger-ui.html`.

**Invoke CLI Updates**

The task for installing the Swagger UI web assets has been updated to simply call `scripts/build.swagger.sh`

```
$ invoke dependencies.install-swagger-ui
2021-04-14 22:49:37,732 [INFO] [tasks.dependencies] Installing Swagger UI assets...
+ echo 'Running the Swagger build within Docker...'
Running the Swagger build within Docker...
+ docker pull node:latest
latest: Pulling from library/node
Digest: sha256:6cbc150709d59d2667f5d34cbf03fb4594dc8b34acb8872f9ab27ba915b28b56
Status: Image is up to date for node:latest
docker.io/library/node:latest
++ pwd
+ docker run --rm -v /Users/jason.parham/code/houston/:/code -w /code node:latest /bin/bash -c './scripts/build.swagger.sh --exec'
+ pushd docs/
/code/docs /code
+ rm -rf node_modules/
+ rm -rf package-lock.json
+ rm -rf swagger_static/
+ mkdir -p swagger_static/
+ npm cache clean -f
npm WARN using --force Recommended protections disabled.
npm notice
npm notice New minor version of npm available! 7.7.6 -> 7.9.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v7.9.0>
npm notice Run `npm install -g npm@7.9.0` to update!
npm notice
+ npm install -g npm@latest

changed 14 packages, and audited 253 packages in 13s

11 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npm notice
npm notice New minor version of npm available! 7.7.6 -> 7.9.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v7.9.0>
npm notice Run `npm install -g npm@7.9.0` to update!
npm notice
+ npm install --legacy-peer-deps

added 2 packages, and audited 3 packages in 4s

found 0 vulnerabilities
+ npm audit fix --force
npm WARN using --force Recommended protections disabled.

up to date, audited 3 packages in 1s

found 0 vulnerabilities
+ cp node_modules/swagger-ui-dist/swagger-ui.css node_modules/swagger-ui-dist/swagger-ui.css.map node_modules/swagger-ui-dist/swagger-ui-bundle.js node_modules/swagger-ui-dist/swagger-ui-es-bundle-core.js node_modules/swagger-ui-dist/swagger-ui-es-bundle.js node_modules/swagger-ui-dist/swagger-ui-standalone-preset.js node_modules/swagger-ui-dist/swagger-ui.js node_modules/swagger-ui-dist/swagger-ui-bundle.js.map node_modules/swagger-ui-dist/swagger-ui-es-bundle-core.js.map node_modules/swagger-ui-dist/swagger-ui-es-bundle.js.map node_modules/swagger-ui-dist/swagger-ui-standalone-preset.js.map node_modules/swagger-ui-dist/swagger-ui.js.map node_modules/swagger-ui-dist/favicon-16x16.png node_modules/swagger-ui-dist/favicon-32x32.png node_modules/swagger-ui-dist/oauth2-redirect.html swagger_static/
+ cp node_modules/typeface-droid-sans/index.css swagger_static/droid-sans.css
+ cp -R node_modules/typeface-droid-sans/files swagger_static/
+ popd
/code
Finished running the build within Docker
+ echo 'Finished running the build within Docker'
+ install
+ set -ex
+ rm -rf app/static/swagger-ui/
+ mkdir -p app/static/swagger-ui/
+ cp -R docs/swagger_static/droid-sans.css docs/swagger_static/favicon-16x16.png docs/swagger_static/favicon-32x32.png docs/swagger_static/files docs/swagger_static/oauth2-redirect.html docs/swagger_static/swagger-ui-bundle.js docs/swagger_static/swagger-ui-bundle.js.map docs/swagger_static/swagger-ui-es-bundle-core.js docs/swagger_static/swagger-ui-es-bundle-core.js.map docs/swagger_static/swagger-ui-es-bundle.js docs/swagger_static/swagger-ui-es-bundle.js.map docs/swagger_static/swagger-ui-standalone-preset.js docs/swagger_static/swagger-ui-standalone-preset.js.map docs/swagger_static/swagger-ui.css docs/swagger_static/swagger-ui.css.map docs/swagger_static/swagger-ui.js docs/swagger_static/swagger-ui.js.map app/static/swagger-ui/
+ exit 0
2021-04-14 22:50:19,124 [INFO] [tasks.dependencies] Swagger UI is installed.
```

**Pull Request Checklist**
- [X] Ensure that the PR is properly formatted
- [X] Ensure that the PR is properly rebased
  - The PR is rebased on `main` (commit: `0d85b6b`)
- [X] Ensure that the PR uses a consolidated database migration
  - _No database migration_
- [x] Ensure that the PR is properly tested
- [x] Ensure that the PR is properly covered
- [X] Ensure that the PR is properly sanitized
- [x] Ensure that the PR is properly reviewed

### CODE REVIEW

**Review by bluemellophone — COMMENTED** [review]
Self-reviewed

**Review by colinwkingen — COMMENTED** [review]
LGTM but can leave approval to Ben or Micheal as requested reviewers. 

What was the error with tempfile.TemporaryDirectory that prompted TemporaryDirectoryGraceful?

**Comment by brmscheiner** [comment]
@bluemellophone when I try to run this in docker-compose I get the following error:

```
houston_1       |   Running setup.py develop for Houston
houston_1       | Successfully installed Houston appdirs-1.4.4 black-20.8b1 brunette-0.2.0 cfgv-3.2.0 codecov-2.1.11 coverage-5.5 distlib-0.3.1 filelock-3.0.12 identify-2.2.3 iniconfig-1.1.1 mock-4.0.3 mypy-extensions-0.4.3 nodeenv-1.6.0 pathspec-0.8.1 pluggy-0.13.1 pre-commit-2.12.0 py-1.10.0 pytest-6.2.3 pytest-cov-2.11.1 pytest-random-order-1.0.4 regex-2021.4.4 swagger-spec-validator-2.7.3 typed-ast-1.4.3 typing-extensions-3.7.4.3 virtualenv-20.4.3 webargs-5.5.3 xdoctest-0.15.4
houston_1       | WARNING: '.' not a valid package name; please use only .-separated package names in setup.py
houston_1       | /usr/local/lib/python3.9/site-packages/setuptools/dist.py:475: UserWarning: The version specified ('0.1.0.b24c6b64') is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
houston_1       |   warnings.warn(
houston_1       | 0.1.0.b24c6b64
houston_1       | 2021-04-15 21:58:40,536 [INFO] [tasks.dependencies] Installing Swagger UI assets...
houston_1       | + echo 'Running the Swagger build within Docker...'
houston_1       | + docker pull node:latest
houston_1       | Running the Swagger build within Docker...
houston_1       | scripts/build.swagger.sh: line 14: docker: command not found
codex_houston_1 exited with code 127
```

**Comment by bluemellophone** [comment]
You cannot run that inside docker-compose because that script relies on Docker being available.  This is a change from before and documented in the PR.

**Review by colinwkingen — APPROVED** [review]
Good to go.

File: tasks/app/run.py [hunk 1]
```diff
@@ -66,6 +67,11 @@ def warmup(
         #         skip_on_failure=True,
         #     )
 
+    if print_routes or app.debug:
+        log.info('Using route rules:')
+        for rule in app.url_map.iter_rules():
+            log.info('\t%r' % (rule,))
+
```
Comment (mmulich) [inline tasks/app/run.py:74]: see also `invoke app.endpoints.list`

Comment (bluemellophone) [reply tasks/app/run.py:74]: Yes, I found this specifically helpful when debugging routes.  I'd argue there is benefit in having both defined.  One is part of the debug printout and another is just a way to on-demand print the endpoints without starting a server.

**Pull request merged**